crates/mysten-network/src/client.rs

```diff
@@ -123,3 +145,122 @@ fn apply_config_to_endpoint(config: &Config, mut endpoint: Endpoint) -> Endpoint
         .initial_connection_window_size(config.http2_initial_connection_window_size)
         .tcp_keepalive(config.tcp_keepalive)
 }
+
+type CacheEntry = (Instant, Vec<SocketAddr>);
+
+/// A caching resolver based on hyper_util GaiResolver
+#[derive(Clone)]
+pub struct CachingResolver {
+    cache: Arc<Mutex<HashMap<Name, CacheEntry>>>,
+}
+
+type SocketAddrs = vec::IntoIter<SocketAddr>;
+
+pub struct CachingFuture {
+    inner: JoinHandle<Result<SocketAddrs, io::Error>>,
+}
+
+impl CachingResolver {
+    pub fn new() -> Self {
+        CachingResolver {
+            cache: Arc::new(Mutex::new(HashMap::new())),
+        }
+    }
+}
+
+impl Default for CachingResolver {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl Service<Name> for CachingResolver {
+    type Response = SocketAddrs;
+    type Error = io::Error;
+    type Future = CachingFuture;
+
+    fn poll_ready(&mut self, _cx: &mut task::Context<'_>) -> Poll<Result<(), io::Error>> {
+        Poll::Ready(Ok(()))
+    }
+
+    fn call(&mut self, name: Name) -> Self::Future {
+        let blocking = {
+            let cache = self.cache.clone();
+            tokio::task::spawn_blocking(move || {
+                let entry = cache.lock().unwrap().get(&name).cloned();
+
+                if let Some((when, addrs)) = entry {
+                    trace!("cached host={:?}", name.as_str());
+
+                    if when.elapsed().as_secs() > 60 {
+                        trace!("refreshing cache for host={:?}", name.as_str());
+                        // Start a new task to update the cache later.
+                        tokio::task::spawn_blocking(move || {
+                            if let Ok(addrs) = (name.as_str(), 0).to_socket_addrs() {
+                                let addrs: Vec<_> = addrs.collect();
+                                trace!("updating cached host={:?}", name.as_str());
+                                cache
+                                    .lock()
+                                    .unwrap()
+                                    .insert(name, (Instant::now(), addrs.clone()));
+                            }
+                        });
+                    }
```

Comment:
Could this lead to multiple tasks trying to update the same entry?

Reply:
Yes, but it didn't seem worth the effort to prevent it from happening.

Reply:
Fair enough!